src/main/java/com/clearspring/analytics/stream/cardinality/RegisterSetStorage.java

```diff
@@ -0,0 +1,129 @@
+package com.clearspring.analytics.stream.cardinality;
+
+import java.lang.reflect.Field;
+
+import sun.misc.Unsafe;
+
+public interface RegisterSetStorage {
+    int get(int bucket);
+    void set(int bucket, int value);
+    int size();
+    int[] readOnlyBits();
+    Type type();
+
+    enum Type {
+        ARRAY_BACKED(1),
+        OFFHEAP(2);
+        public int type;
+
+        Type(int type) {
+            this.type = type;
+        }
+
+        public static Type fromInt(int storageType) {
+            for (Type t : values())
+                if (t.type == storageType)
+                    return t;
+            throw new RuntimeException("Unknown storage type: "+storageType);
+        }
+    }
+
+    class ArrayBackedRegisterSetStorage implements RegisterSetStorage {
```

Comment:
Don't make inner or nested classes in general, please. Especially in ambiguous cases like these where people might not even reslize they're static.